src/group_theory/index.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2021 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+-/
+
+import group_theory.coset
+import set_theory.cardinal
+
+/-!
+# Index of a Subgroup
+
+In this file we define the index of a subgroup, and prove several divisibility properties.
+
+## Main definitions
+
+- `H.index` : the index of `H : subgroup G`
+
+# Main results
+
+- `index_mul_card` : `H.index * fintype.card H = fintype.card G`
+- `index_dvd_card` : `H.index ∣ fintype.card G`
+- `index_eq_mul_of_le` : If `H ≤ K`, then `H.index = K.index * (H.subgroup_of K).index`
+- `index_dvd_of_le` : If `H ≤ K`, then `K.index ∣ H.index`
+
+-/
+
+namespace subgroup
+
+variables {G : Type*} [group G] (H : subgroup G)
+
+/-- The index of a subgroup as a natural number -/
+noncomputable def index : ℕ :=
+(cardinal.mk (quotient_group.quotient H)).to_nat
+
+lemma index_eq_card [fintype (quotient_group.quotient H)] :
+  H.index = fintype.card (quotient_group.quotient H) :=
+cardinal.mk_to_nat_eq_card
+
+lemma index_mul_card [fintype G] [fintype H] :
```

Comment:
```suggestion
lemma index_mul_card [fintype G] {hH : fintype H} :
```
This way unification instead of typeclass inference is used. Would that be better?
Isn't there an instance `fintype G -> fintype (H : subgroup G)`?

Reply:
Such an instance exists, but it requires some decidable hypothesis, I think.

Reply:
I see, well both `[]` and `{}` have their pros and cons. Maybe leave the `[]` for now.